sdks/java/io/hive/src/main/java/org/apache/beam/sdk/io/hive/HiveIO.java

```diff
@@ -0,0 +1,474 @@
+      IMetaStoreClient client = HCatUtil.getHiveMetastoreClient(hiveConf);
+      Table table = HCatUtil.getTable(client, spec.getDatabase(),
+        spec.getTable());
+      client.close();
+      return StatsUtils.getFileSizeForTable(hiveConf, table);
+    }
+
+    @Override
+    public List<BoundedSource<HCatRecord>> split(long desiredBundleSizeBytes,
+                                                PipelineOptions options) {
+      //TODO: We are really not using the bundle size at run time
+      //to make the split
+      //just going with the default splits returned by Hive
+      //unable to explicitly control the splits returned by Hive
+      List<BoundedSource<HCatRecord>> sources = new ArrayList<>();
+      ReaderContext cntxt = getReaderContext();
+      for (int i = 0; i < cntxt.numSplits(); i++) {
+        //create an instance of BoundedHiveSource for each of the split
+        //to be read in parallel
+        sources.add(new BoundedHiveSource(spec.withSplitId(i).withContext(cntxt)));
+      }
+      return sources;
+    }
+
+    private ReaderContext getReaderContext() {
+      ReadEntity entity = new ReadEntity.Builder().withDatabase(spec.getDatabase())
+          .withTable(spec.getTable()).withFilter(spec.getFilter()).build();
+      Map<String, String> hiveConfig = new HashMap<String, String>();
+      hiveConfig.put(HCatConstants.HCAT_METASTORE_URI, spec.getMetastoreUri());
+      HCatReader masterReader = DataTransferFactory.getHCatReader(entity, hiveConfig);
+      ReaderContext cntxt = null;
+      try {
+        cntxt = masterReader.prepareRead();
+        } catch (HCatException e) {
+        LOG.error("Exception while preparing for read", e);
+      }
+      return cntxt;
+    }
+
+    private static class BoundedHiveReader extends BoundedSource.BoundedReader<HCatRecord> {
+      private final BoundedHiveSource source;
+      private HCatRecord current;
+      Iterator<HCatRecord> hcatIterator;
+
+      public BoundedHiveReader(BoundedHiveSource boundedHiveSource) {
+        this.source = boundedHiveSource;
+      }
+
+      @Override
+      public boolean start() {
+        try {
+          HCatReader reader = DataTransferFactory.getHCatReader(source.spec.getContext(),
+              source.spec.getSplitId());
+          hcatIterator = reader.read();
+        } catch (HCatException e) {
+          LOG.error("Exception while reading", e);
+          return false;
+        }
+        return advance();
+      }
+
+      @Override
+      public boolean advance() {
+        if (hcatIterator.hasNext()) {
+          current = hcatIterator.next();
+          return true;
+        } else {
+          return false;
+        }
+      }
+
+      @Override
+      public BoundedHiveSource getCurrentSource() {
+        return source;
+      }
+
+      @Override
+      public HCatRecord getCurrent() {
+        return current;
+      }
+
+      @Override
+      public void close() {
+        //nothing to close/release
```

Comment:
Interesting... I'm surprised that reading a result set does not involve any closeable resources.